backend/queue/queue.go

```diff
@@ -20,34 +21,68 @@ var (
 	backendIDKeyPrefix = store.NewKeyBuilder("backends").Build()
 )
 
-type BackendIDGetter interface {
-	GetBackendID() int64
+// Client interface to queue backend
+type Client interface {
+	// Enqueue Item
+	Enqueue(context.Context, Item) error
+	// Reserve an Item from the queue
+	// Blocks until an item can be reserved or the context expires
+	Reserve(ctx context.Context, queue string) (Reservation, error)
 }
 
-// Item is a Queue item.
+// Item is a Queue Item.
 type Item struct {
-	key   string
-	value string
+	// ID of queue item. Ignored on Enqueue
+	ID string
+	// Queue name
+	Queue string
+	// Value of queue item
+	Value []byte
 }
 
-// Key returns the key of the Item.
-func (i *Item) Key() string {
-	return i.key
+// Reservation for a Queue Item.
+// Must be either Ack'd or Nack'd.
+type Reservation interface {
+	Item() Item
+	// Ack the Item was handeled and can be deleted.
+	Ack(context.Context) error
+	// Nack the Item could not be handled. Return to the queue.
+	Nack(context.Context) error
 }
 
-// Value returns the value of the Item.
-func (i *Item) Value() string {
-	return i.value
+// ClusteredQueue is a Client wrapper meant to function as a simple pub/sub
+// across all active sensu backends in a cluster.
+// Given a queueName, ClusteredQueue Enqueues to queueName/{{backendID}} for
+// each backend, and Reserves only from queueName/{{this backendID}}
+type ClusteredQueue struct {
+	// Client underlying queue implementation
+	Client Client
+	// GetBackendID gets the current backend process's id.
+	// TODO(ck): just hostname per backend/opc.go?
```

Comment:
I think we should move to backend IDs being hostnames, or optionally named by config file. If backends share the same hostname, that's on them and sensu execution will be weird. This is what agents do, so no reason for backends not to.